server/src/main/java/org/elasticsearch/transport/TcpTransportKeepAlive.java

```diff
@@ -0,0 +1,204 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.transport;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.component.Lifecycle;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.metrics.CounterMetric;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.AbstractLifecycleRunnable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class TcpTransportKeepAlive implements Closeable {
+
+    static final int PING_DATA_SIZE = -1;
+
+    private final Logger logger = LogManager.getLogger(TcpTransportKeepAlive.class);
+    private final CounterMetric successfulPings = new CounterMetric();
+    private final CounterMetric failedPings = new CounterMetric();
+    private final ConcurrentMap<TimeValue, ScheduledPing> pingIntervals = ConcurrentCollections.newConcurrentMap();
+    private final Lifecycle lifecycle = new Lifecycle();
+    private final ThreadPool threadPool;
+    private final PingSender pingSender;
+    private final BytesReference pingMessage;
+
+    TcpTransportKeepAlive(ThreadPool threadPool, PingSender pingSender) {
+        this.threadPool = threadPool;
+        this.pingSender = pingSender;
+
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.writeByte((byte) 'E');
+            out.writeByte((byte) 'S');
+            out.writeInt(PING_DATA_SIZE);
+            pingMessage = out.bytes();
+        } catch (IOException e) {
+            throw new AssertionError(e.getMessage(), e); // won't happen
+        }
+
+        this.lifecycle.moveToStarted();
+    }
+
+    void registerNodeConnection(List<TcpChannel> nodeChannels, ConnectionProfile connectionProfile) {
+        TimeValue pingInterval = connectionProfile.getPingInterval();
+        if (pingInterval.millis() < 0) {
+            return;
+        }
+
+        final ScheduledPing scheduledPing = pingIntervals.computeIfAbsent(pingInterval, ScheduledPing::new);
+        scheduledPing.ensureStarted();
+
+        for (TcpChannel channel : nodeChannels) {
+            scheduledPing.addChannel(channel);
```

Comment:
I wonder if we should first register the close listener? is it guaranteed that we execute the listener if the channel is already closed? I guess it is, I don't remember.